### PR TITLE
Resource quota changes to give memory back sooner to total quota

### PR DESCRIPTION
**Pull request by Vignesh2208**
Copying craig's changes to give memory back to total quota sooner.

@veblush

### CODE REVIEW

File: src/core/lib/resource_quota/memory_quota.h [hunk 1]
```diff
@@ -323,6 +329,27 @@ class GrpcMemoryAllocatorImpl final : public EventEngineMemoryAllocatorImpl {
  private:
   // Primitive reservation function.
   absl::optional<size_t> TryReserve(MemoryRequest request) GRPC_MUST_USE_RESULT;
+  void MaybeDonateBack() {
```
Comment (ananda1066) [inline src/core/lib/resource_quota/memory_quota.h:332]: Please comment this function.

Comment (ctiller) [reply src/core/lib/resource_quota/memory_quota.h:332]: And move it to the .cc

File: src/core/lib/resource_quota/memory_quota.h [hunk 2]
```diff
@@ -295,7 +297,11 @@ class GrpcMemoryAllocatorImpl final : public EventEngineMemoryAllocatorImpl {
     // Add the released memory to our free bytes counter... if this increases
     // from  0 to non-zero, then we have more to do, otherwise, we're actually
     // done.
-    if (free_bytes_.fetch_add(n, std::memory_order_release) != 0) return;
+    size_t prev_free = free_bytes_.fetch_add(n, std::memory_order_release);
+    if (prev_free + n > 1024 * 1024) {
```
Comment (ctiller) [inline src/core/lib/resource_quota/memory_quota.h:301]: Let's make a symbolic constant for 1024*1024... kMaxQuotaBufferSize?

Comment (Vignesh2208) [reply src/core/lib/resource_quota/memory_quota.h]: Done.

File: src/core/lib/resource_quota/memory_quota.h (hunk 1 above)
Comment (Vignesh2208) [reply src/core/lib/resource_quota/memory_quota.h]: Done.

File: src/core/lib/resource_quota/memory_quota.cc [hunk 3]
```diff
@@ -376,8 +400,13 @@ void BasicMemoryQuota::Start() {
...
         if (GRPC_TRACE_FLAG_ENABLED(grpc_resource_quota_trace)) {
-          gpr_log(GPR_INFO, "RQ: %s perform %s reclamation",
-                  self->name_.c_str(), std::get<0>(arg));
+          double free = self->free_bytes_.load();
+          if (free < 0) free = 0;
+          size_t quota_size = self->quota_size_.load();
+          gpr_log(GPR_INFO,
+                  "RQ: %s perform %s reclamation. Available free bytes: %f, "
+                  "total quota_size: %zu",
+                  self->name_.c_str(), std::get<0>(arg), free, quota_size);
```
Comment (ananda1066) [inline src/core/lib/resource_quota/memory_quota.cc:409]: could we simplify to something like std::max(0, free)? or grpc::clamp?

Comment (ananda1066) [inline src/core/lib/resource_quota/memory_quota.cc:409]: could we simplify to something like std::max(0, free)? or grpc::clamp?

Comment (Vignesh2208) [reply src/core/lib/resource_quota/memory_quota.cc:408]: Done

**Comment by ctiller** [comment]
Probably prefer max(size_t(0), free_bytes.load())?

reinterpret_cast seems OP

**Comment by Vignesh2208** [comment]
> Probably prefer max(size_t(0), free_bytes.load())?
> 
> reinterpret_cast seems OP

This didn't work on linux. Had to cast the 0 to inptr_t to make it happy :)

**Review by ananda1066 — APPROVED** [review]

**Pull request merged**